Consul Upstream Load Balancing lab (#244)
hashicorp/katakoda: consul-upstream-load-balancing/step2.md

> Once verified you can access the _backend_ service and that the *round_robin* policy is applied you can apply new policies for the load balancing and verify how these affect the requests' resolution.

Reviewer: ```suggestion
Once verified, you can access the _backend_ service and that the *round_robin* policy is applied you can apply new policies for the load balancing and verify how these affect the requests' resolution.
```

### Configure service defaults

> In order to enable service resolution and apply load balancer policies you need to define the service protocol in a `service-defaults` configuration entry.

Reviewer: ```suggestion
In order to enable service resolution and apply load balancer policies, you need to define the service protocol in a `service-defaults` configuration entry.
```

Name           = "backend"
Protocol       = "http"
```

`docker exec server consul config write /etc/consul.d/default.hcl`{{execute}}

Example output:

```
Config entry written: service-defaults/backend
```

### Configure service resolution with sticky sessions

A common requirements for many applications is to have the possibility to redirect all the requests from a specific client to the same server.

You can achieve this configuration using the `maglev` policy provided in the `hash-resolver.hcl`{{open}} file:

```
Kind           = "service-resolver"
Name           = "backend"
LoadBalancer = {
  EnvoyConfig = {
    Policy = "maglev"
    HashPolicies = [
      {
        Field = "header"
        FieldValue = "x-user-id"
      }
    ]
  }
}
```

This configuration creates a `service-resolver` configuration, for the service `backend` that uses the content of the `x-user-id` header to resolve requests.

You can apply the policy using the `consul config` command.

`docker exec server consul config write /etc/consul.d/hash-resolver.hcl`{{execute}}

Example output:

```
Config entry written: service-resolver/backend
```

### Verify the policy is applied

> Once the policy is in place you can test it using the `curl` command and applying the `x-user-id` header to the request:

Reviewer: ```suggestion
Once the policy is in place, you can test it using the `curl` command and applying the `x-user-id` header to the request:
```

`docker exec client curl -s localhost:9192 -H "x-user-id: 12345"`{{execute}}

Example output:

```
{
  "name": "main",
  "uri": "/",
  "type": "HTTP",
  "ip_addresses": [
    "172.18.0.4"
  ],
  "start_time": "2020-10-01T16:15:47.950151",
  "end_time": "2020-10-01T16:15:47.950581",
  "duration": "430.088µs",
  "body": "Hello World",
  "code": 200
}
```

> If you execute the previous command multiple times you will always be redirected to the same instance of the _backend_ service.

Reviewer: ```suggestion
Execute the curl command multiple times, you will always be redirected to the same instance of the _backend_ service.

`docker exec client curl -s localhost:9192 -H "x-user-id: 12345"`{{execute}}
```

> ### Check configuration

Reviewer: This is great!

Once verified you can access the _backend_ service and that the _random_ policy is applied you can apply new policies for the load balancing and verify how these affect the requests' resolution.

### Configure service defaults

In order to enable service resolution and apply load balancer policies you need to define the service protocol as a `service-defaults` entry.

The lab provides a configuration file for this, `default.hcl`{{open}} that you can use to apply the configuration.

```
Kind           = "service-defaults"
Name           = "backend"
Protocol       = "http"
```

`docker exec server consul config write /etc/consul.d/default.hcl`{{execute}}

Example output:

```
Config entry written: service-defaults/backend
```

### Configure a sticky session for service resolution

> A common requirements for many applications is to have the possibility to redirect all the requests from a specific client to the same server.

Author: For the first point that seems a good feature to have, I mean better reconstruct the "state" than deny requests. 
Do we have some more logic on that? I mean my understanding is that the extra component (in this case the header) was used to generate one hash that would remain consistent. 

I might need help in phrasing this but I think we can pass the message that "sticky session is consistent (reliable/predictable) given a stable service configuration, if the number of services we load balance on changes, some sessions might be routed to a different host after the change but will then maintain the new association."

> A common requirements for many applications is to have the possibility to redirect all the requests from a specific client to the same server.

Reviewer 2: that's mostly accurate I would make a couple edits:

> sticky sessions are consistent (reliable/predictable) given a stable service configuration. If the number of backend hosts changes, a fraction of the sessions will be routed to a new host after the change.

If the number of backend services changes, the routing *will* change for some services.

> A common requirements for many applications is to have the possibility to redirect all the requests from a specific client to the same server.

Author: I am adding this later on in the

    HashPolicies = [
      {
        Field = "header"
        FieldValue = "x-user-id"
      }
    ]
  }
}
```

This configuration creates a `service-resolver` configuration, for the service `backend` that uses the content of the `x-user-id` header to resolve requests.

You can apply the policy using the `consul config` command.

`docker exec server consul config write /etc/consul.d/hash-resolver.hcl`{{execute}}

Example output:

```
Config entry written: service-resolver/backend
```

### Verify the policy is applied

Once the policy is in place, you can test it using the `curl` command and applying the `x-user-id` header to the request:

`docker exec client curl -s localhost:9192 -H "x-user-id: 12345"`{{execute}}

Example output:

```
{
  "name": "main",
  "uri": "/",
  "type": "HTTP",
  "ip_addresses": [
    "172.18.0.4"
  ],
  "start_time": "2020-10-01T16:15:47.950151",
  "end_time": "2020-10-01T16:15:47.950581",
  "duration": "430.088µs",
  "body": "Hello World",
  "code": 200
}
```

Execute the curl command multiple times, you will always be redirected to the same instance of the _backend_ service.

`docker exec client curl -s localhost:9192 -H "x-user-id: 12345"`{{execute}}

> 

Author: ```suggestion

<div style="background-color:#fcf6ea; color:#866d42; border:1px solid #f8ebcf; padding:1em; border-radius:3px; margin:24px 0;">
  <p><strong>Note:</strong> sticky sessions are consistent given a stable service configuration. If the number of backend hosts changes, a fraction of the sessions will be routed to a new host after the change.
</p></div>
```
